### ISSUE TITLE
Hunts API very slow after 3.1.0.2 upgrade

### ISSUE THREAD

**Issue by huskyjag** [google/grr#466]
Hi there

We've been upgrading our GRR deployment from 3.0.7 to 3.1.0.2.post2. After the upgrade, we noticed that the hunts API (`/api/hunts/<hunt_id>`) is very slow. Querying the API for hunts that were created **_before_** the upgrade are fast. The call seems to hang at `HuntRunner.GetClientsByStatus` (https://github.com/google/grr/blob/server-3.1.0.2/grr/lib/hunts/implementation.py#L1010).

Running `GetClientsByStatus` body in the `grr_console` confirms this assumption (reduced verbose output for readability).

`OLD_HUNT` was created before the upgrade, query completes within one second:
```python
In [2]: fast_hunt_query = aff4.FACTORY.Open(HUNTS_ROOT_PATH.Add("H:OLD_HUNT"), aff4_type=hunts.GRRHunt, token=token)
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts'], 'r'), {}) with reason

In [3]: stat = fast_hunt_query.GetClientsByStatus()
ata_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/Stream'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000008'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000009'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000000'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000001'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000002'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000003'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000004'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000005'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000006'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/AllClients/UnversionedStream/0000000007'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/Stream'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000008'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000009'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000000'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000001'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000002'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000003'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000004'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000005'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000006'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:OLD_HUNT/CompletedClients/UnversionedStream/0000000007'], 'r'), {}) with reason:
```

`NEW_HUNT` was created after the upgrade. Query takes around 30 seconds. Note: the query limit for `NEW_HUNT` was set to 500.
```python
In[3]: slow_hunt_query = aff4.FACTORY.Open(HUNTS_ROOT_PATH.Add("H:NEW_HUNT"), aff4_type=hunts.GRRHunt, token=token)
ata_store_access GRANTED [...] to root (([u'aff4:/hunts/H:NEW_HUNT'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:NEW_HUNT/AllClients'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([<aff4:/hunts/H:NEW_HUNT/AllClients age=1970-01-01 00:00:00>], 'rq'), {}) with reason:
data_store_access GRANTED [...] to root (([<aff4:/hunts/H:NEW_HUNT/AllClients/Results age=2017-02-03 08:44:56>], 'qr'), {}) with reason:
# --> 20 seconds spent here <--
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:NEW_HUNT/CompletedClients'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts/H:NEW_HUNT'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([u'aff4:/hunts'], 'r'), {}) with reason:
data_store_access GRANTED [...] to root (([<aff4:/hunts/H:NEW_HUNT/CompletedClients age=1970-01-01 00:00:00>], 'rq'), {}) with reason:
data_store_access GRANTED [...] to root (([<aff4:/hunts/H:NEW_HUNT/CompletedClients/Results age=2017-02-03 08:45:12>], 'qr'), {}) with reason:
```

Around 20 seconds are spend between querying `<aff4:/hunts/H:NEW_HUNT/AllClients/Results age=2017-02-03 08:44:56>` and `aff4:/hunts/H:NEW_HUNT/CompletedClients`.

MySQLs slow query log reveils the resulting query. It appears to **scan through all rows** and actually returns more clients than in the hunt included.
```sql
SET timestamp=1486114226;
SELECT aff4.value, aff4.timestamp, subjects.subject
FROM aff4
JOIN subjects ON aff4.subject_hash=subjects.hash
JOIN (
                SELECT subject_hash, MAX(timestamp) timestamp
                FROM aff4
                JOIN subjects ON aff4.subject_hash=subjects.hash
                WHERE aff4.attribute_hash=unhex(md5('aff4:sequential_value'))
                AND subjects.subject like 'aff4:/hunts/H:NEW_HUNT/AllClients/Results/%'
                AND subjects.subject > 'aff4:/hunts/H:NEW_HUNT/AllClients/Results/0000000000000000.-00001'
                GROUP BY subject_hash
     ) maxtime ON aff4.subject_hash=maxtime.subject_hash
AND aff4.timestamp=maxtime.timestamp
WHERE aff4.attribute_hash=unhex(md5('aff4:sequential_value'));

SET timestamp=1486114242;
SELECT aff4.value, aff4.timestamp, subjects.subject
FROM aff4
JOIN subjects ON aff4.subject_hash=subjects.hash
JOIN (
                SELECT subject_hash, MAX(timestamp) timestamp
                FROM aff4
                JOIN subjects ON aff4.subject_hash=subjects.hash
                WHERE aff4.attribute_hash=unhex(md5('aff4:sequential_value'))
                AND subjects.subject like 'aff4:/hunts/H:NEW_HUNT/CompletedClients/Results/%'
                AND subjects.subject > 'aff4:/hunts/H:NEW_HUNT/CompletedClients/Results/0000000000000000.-00001'
                GROUP BY subject_hash
     ) maxtime ON aff4.subject_hash=maxtime.subject_hash
AND aff4.timestamp=maxtime.timestamp
WHERE aff4.attribute_hash=unhex(md5('aff4:sequential_value'))
```

Has anyone else experienced this issue? Am I missing something? Thanks! :)

**Comment by pidydx** [comment]
There were changes that started using "like" against the subjects table and the subjects table uses a very long key.  I had thought that ScanAttributes was the only place that did this and that key could be substantially reduced in length, but this query appears to be have generated differently so I will have to look into it.

TLDR; Some functionality was added that requires using more "like" queries that are extremely slow and it needs to be reexamined.

**Comment by huskyjag** [comment]
@pidydx Great, thanks! Let me know if I can help you anyhow.

**Comment by pidydx** [comment]
Roughly how many clients do you have deployed?

**Comment by huskyjag** [comment]
@pidydx Sorry for my late response - been spending some offline time.

We've deployed GRR on roughly 20k clients.

**Comment by huskyjag** [comment]
Hi there. Has there been any progress on this?

Also, will this even be relevant in the future as GRR is going to use a completely new Model (using Google Spanner)?

Thanks :)

**Comment by grrrrrrrrr** [comment]
Hey, there hasn't been any progress, sorry.

You are right, this problem will completely disappear once we switch to the new database schema.

In the meantime, this query could probably be optimized *a lot* as a temporary fix until the new tech is ready but this is quite hard for us to do since we don't have a database big enough that it shows this behavior.

I'll see if we can speed up the conversion of collections to the new schema which is the core of your problem here.

**Comment by pidydx** [comment]
I have a working solution for this that I am testing, but it involves some significant schema changes to the MySQL datastore.  I am hoping to get this code back to the project soon.

**Comment by huskyjag** [comment]
@grrrrrrrrr @pidydx Thanks for the update. Since we have a pretty big MySQL database, we'd be happy to test these changes. Just let me know.

Any plans on the release of the new model? We're quite curious about it :-)

**Comment by pidydx** [comment]
If you want to shoot me an email or connect on the google groups we can chat a little about that.

**Comment by huskyjag** [comment]
@pidydx Thanks! Dropped you a PM on the google group.

**Comment by grrrrrrrrr** [comment]
New database schema is live, this is obsolete.
